### ISSUE TITLE
Connections to the DW won't (always) recreate if connected via SSH tunnel

### ISSUE THREAD

**Issue by paoliniluis** [metabase/metabase#23452]
**Describe the bug**
If Metabase connects to a DW via an SSH tunnel and a connection gets killed in the DB, the connection won't be restored till the entire connection pool is recreated. This is a very strange behavior as it doesn't happen when connected directly to the DW without the SSH tunnel.

**Logs**
Logs say that there are 2 connections when there's only one! (that's for another issue)

**To Reproduce**
1) Set up Metabase (use a MB_JDBC_DATA_WAREHOUSE_MAX_CONNECTION_POOL_SIZE=2 to facilitate the repro) + an SSH tunnel + a DB (you can use https://github.com/paoliniluis/mariadb-10.1.48-ssh which is the ssh tunnel and a mysql db with the sample database)
2) connect your Metabase to the database via SSH, not directly
3) build 4 questions based on this connection, anything could be
4) add these 4 questions to a dashboard and fire it up. Metabase will use the 2 connections available
.... fine for now
5) now get into the MariaDB and kill the youngest connection (you have to run "show full processlist;" and then pick the Metabase connection with the lowest ID, and then use "kill xx" being xx the number of the ID). Run "show full processlist" again to check that Metabase has only 1 connection
6) Now run your dashboard again a few times, and go to the DB and run "show full processlist". See that the connection was never recreated.

(you might have to kill the connection a few times, I found some cases where it does recreate the connection, but mostly doesn't)

**Expected behavior**
Recreate connections when possible!

**Screenshots**
![image](https://user-images.githubusercontent.com/1711649/174694573-470805ea-0733-495c-b1e1-caf403b0854f.png)
![image](https://user-images.githubusercontent.com/1711649/174694593-bd5b03bd-aae7-400d-94dc-70330ea62724.png)

**Information about your Metabase Installation:**
- Your browser and the version: Brave latest
- Your operating system: Pop OS 22.04
- Your databases: mariadb:10.1.48 and postgres 12 (both with sample database)
- Metabase version: 43.2
- Metabase hosting environment: Docker
- Metabase internal database: Postgres 14.3

**Severity**
P1'ish?

**Additional context**
If you kill the connection with the highest ID, it will be recreated 🤷🏻 . Also if you don't use the SSH tunnel, it will be recreated as well

Same thing happens with postgres
EDIT: commands for postgres are 
SELECT * FROM pg_stat_activity;
select pg_terminate_backend(xx)

**Comment by dpsutton** [comment]
I'm unable to reproduce this.

I've got a dashboard with 5 separate questions on it:
<img width="1397" alt="image" src="https://user-images.githubusercontent.com/6377293/175665395-e28c56da-ab63-4c23-8b9e-cebf51f115be.png">
I've set my pool size to 2

and then I kill the earlier of the two connections:
```sql
mysql> show full processlist;
+----+-----------------+-----------------+--------+---------+--------+------------------------+-----------------------+
| Id | User            | Host            | db     | Command | Time   | State                  | Info                  |
+----+-----------------+-----------------+--------+---------+--------+------------------------+-----------------------+
|  5 | event_scheduler | localhost       | NULL   | Daemon  | 134627 | Waiting on empty queue | NULL                  |
|  8 | dan             | localhost       | racing | Query   |      0 | init                   | show full processlist |
| 33 | dan             | localhost:51799 | racing | Sleep   |      7 |                        | NULL                  |
| 34 | dan             | localhost:51801 | racing | Sleep   |      7 |                        | NULL                  |
+----+-----------------+-----------------+--------+---------+--------+------------------------+-----------------------+
4 rows in set (0.00 sec)

mysql> kill 33;
Query OK, 0 rows affected (0.00 sec)

mysql> show full processlist;
+----+-----------------+-----------------+--------+---------+--------+------------------------+-----------------------+
| Id | User            | Host            | db     | Command | Time   | State                  | Info                  |
+----+-----------------+-----------------+--------+---------+--------+------------------------+-----------------------+
|  5 | event_scheduler | localhost       | NULL   | Daemon  | 134699 | Waiting on empty queue | NULL                  |
|  8 | dan             | localhost       | racing | Query   |      0 | init                   | show full processlist |
| 34 | dan             | localhost:51801 | racing | Sleep   |     79 |                        | NULL                  |
+----+-----------------+-----------------+--------+---------+--------+------------------------+-----------------------+
3 rows in set (0.00 sec)
```

I get an error (visible in the screenshot)
> cause | "Timed out after 3.0 mins."
 
But then everything keeps working. The dashboard even automatically refreshes itself.

The timeout seems to happen with our connection pool in checking if its broken:
> "com.mchange.v2.resourcepool.BasicResourcePool.ensureNotBroken(BasicResourcePool.java:1824)"

As far as I can tell, everything is working correctly. There's a default timeout of 3 minutes which makes this feel terrible, but that's the timelimit we set for "wait and see if something happens".

I then added `socketTimeout=2000` (2000 milliseconds)
<img width="663" alt="image" src="https://user-images.githubusercontent.com/6377293/175666869-70e3de83-1c09-44d2-be5f-b9ff4d32fecd.png">
and this issue goes away completely. I'm not arguing that `2000` is a good value for that, just that we can tweak the concept of how long to wait on a connection before giving up on it. All in all I think everything is working correctly here.

**Comment by dpsutton** [comment]
Further investigations:

Using `networkTimeout=2000` means all queries lasting that long will break:
<img width="857" alt="image" src="https://user-images.githubusercontent.com/6377293/175667464-3c475104-ca84-4672-9937-440abc330ea3.png">

But using `connectTimeout=2000` means long running queries are fine, but we abandon connections if we cannot establish a connection within 2 seconds

<img width="555" alt="image" src="https://user-images.githubusercontent.com/6377293/175667590-c23aaa20-02fb-4846-a037-118284460989.png">

**Comment by paoliniluis** [comment]
I think we can close this @dpsutton , what do you think?

**Comment by dpsutton** [comment]
@paoliniluis agree. We can always reopen if we get reports. But as far as we know we are good.
